Bottom: edbdea4549c701c467994033715965a4c5547615
Top:    ea15f1a38ff89a83495ba2ae54a702f962a2eb45
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-11-14 16:54:39 +0800

Refresh of tuntap-ndo_xdp_xmit-support

---

diff --git a/drivers/vhost/net.c b/drivers/vhost/net.c
index cee170f..6b9140a 100644
--- a/drivers/vhost/net.c
+++ b/drivers/vhost/net.c
@@ -175,6 +175,16 @@ static void vhost_net_buf_unproduce(struct vhost_net_virtqueue *nvq)
 	}
 }
 
+static int vhost_net_buf_peek_len(void *ptr)
+{
+	if (ptr & TUN_XDP_FLAG) {
+		struct xdp_buff *xdp = ptr & ~TUN_XDP_FLAG;
+		return xdp->data_end - xdp->data;
+	}
+
+	return __skb_array_len_with_tag(ptr);
+}
+
 static int vhost_net_buf_peek(struct vhost_net_virtqueue *nvq)
 {
 	struct vhost_net_buf *rxq = &nvq->rxq;
@@ -186,7 +196,7 @@ static int vhost_net_buf_peek(struct vhost_net_virtqueue *nvq)
 		return 0;
 
 out:
-	return tun_xdp_peek_len(vhost_net_buf_get_ptr(rxq));
+	return vhost_net_buf_peek_len(vhost_net_buf_get_ptr(rxq));
 }
 
 static void vhost_net_buf_init(struct vhost_net_buf *rxq)
